src/Forms/GridField/GridFieldVersionTag.php

```diff
@@ -0,0 +1,186 @@
+    public function getVersionedLabelFields()
+    {
+        return $this->versionedLabelFields;
+    }
+
+    /**
+     * @param array $versionedLabelFields
+     * @return VersionedGridFieldState
+     */
+    public function setVersionedLabelFields($versionedLabelFields)
+    {
+        $this->versionedLabelFields = $versionedLabelFields;
+        return $this;
+    }
+
+    /**
+     * Modify the list of columns displayed in the table.
+     *
+     * @see {@link GridFieldDataColumns->getDisplayFields()}
+     * @see {@link GridFieldDataColumns}.
+     *
+     * @param GridField $gridField
+     * @param array $columns List reference of all column names.
+     */
+    public function augmentColumns($gridField, &$columns)
+    {
+        // Skip if not versioned, or column already set
+        if ($this->getColumn()) {
+            return;
+        }
+
+        $matchedVersionedFields = array_intersect(
+            $columns ?? [],
+            $this->versionedLabelFields
+        );
+
+        if (count($matchedVersionedFields ?? []) > 0) {
+            // Get first matched column
+            $this->setColumn(reset($matchedVersionedFields));
+        } elseif ($columns) {
+            // Use first column if none of preferred matches
+            $this->setColumn(reset($columns));
+        }
+    }
+
+    /**
+     * Names of all columns which are affected by this component.
+     *
+     * @param GridField $gridField
+     * @return array
+     */
+    public function getColumnsHandled($gridField)
+    {
+        return $this->getColumn() ? [$this->getColumn()] : [];
+    }
+
+    /**
+     * HTML for the column, content of the <td> element.
+     *
+     * @param GridField $gridField
+     * @param DataObject $record Record displayed in this row
+     * @param string $columnName
+     * @return string HTML for the column. Return NULL to skip.
+     */
+    public function getColumnContent($gridField, $record, $columnName)
+    {
+        $flagContent = '';
+        $flags = $this->getStatusFlags($record);
+        foreach ($flags as $class => $data) {
+            $flagAttributes = [
+                'class' => "ss-gridfield-badge badge status-{$class}",
+            ];
+            if (isset($data['title'])) {
+                $flagAttributes['title'] = $data['title'];
+            }
+            $flagContent .= ' ' . HTML::createTag('span', $flagAttributes, Convert::raw2xml($data['text']));
+        }
+        return $flagContent;
+    }
+
+    /**
+     * Attributes for the column
+     */
+    public function getColumnAttributes($gridField, $record, $columnName)
+    {
+        return [];
+    }
+
+    /**
+     * Metadata for the column
+     */
+    public function getColumnMetadata($gridField, $columnName)
+    {
+        return [];
+    }
+
+    /**
+     * Get status flags for a given record
+     */
+    private function getStatusFlags(DataObject $record): array
+    {
+        if ($record->hasExtension(Versioned::class)) {
```

Comment:
This check that this class does not have the Versioned extension is needed in order, firstly, not to have a duplicate badge and, secondly, not to override the badge of the current state of the current object.

Reply:
So there is already a versioned badge, and this PR is adding a new one?  Is the original status badge wrong?  Where is the code for the original status badge. Would it make sense to move this code there, or at least deprecate the old code since it's presumably wrong?